src/app/SimpleSubscriptionResumptionStorage.cpp

```diff
@@ -0,0 +1,498 @@
+        {
+            ReturnErrorOnFailure(writer.Put(kEndpointIdTag, subscriptions.mSubscriptions[i].mAttributePaths[j].mEndpointId));
+            ReturnErrorOnFailure(writer.Put(kClusterIdTag, subscriptions.mSubscriptions[i].mAttributePaths[j].mClusterId));
+            ReturnErrorOnFailure(writer.Put(kAttributeIdTag, subscriptions.mSubscriptions[i].mAttributePaths[j].mAttributeId));
+        }
+
+        ReturnErrorOnFailure(
+            writer.Put(kPathCountTag, static_cast<uint16_t>(subscriptions.mSubscriptions[i].mEventPaths.AllocatedCount())));
+        for (size_t j = 0; j < subscriptions.mSubscriptions[i].mEventPaths.AllocatedCount(); j++)
+        {
+            if (subscriptions.mSubscriptions[i].mEventPaths[j].mIsUrgentEvent)
+            {
+                ReturnErrorOnFailure(writer.Put(kPathTypeTag, EventPathType::kUrgent));
+            }
+            else
+            {
+                ReturnErrorOnFailure(writer.Put(kPathTypeTag, EventPathType::kNonUrgent));
+            }
+            ReturnErrorOnFailure(writer.Put(kEndpointIdTag, subscriptions.mSubscriptions[i].mEventPaths[j].mEndpointId));
+            ReturnErrorOnFailure(writer.Put(kClusterIdTag, subscriptions.mSubscriptions[i].mEventPaths[j].mClusterId));
+            ReturnErrorOnFailure(writer.Put(kAttributeIdTag, subscriptions.mSubscriptions[i].mEventPaths[j].mEventId));
+        }
+
+        ReturnErrorOnFailure(writer.EndContainer(subscriptionContainerType));
+    }
+
+    ReturnErrorOnFailure(writer.EndContainer(arrayType));
+
+    const auto len = writer.GetLengthWritten();
+    VerifyOrReturnError(CanCastTo<uint16_t>(len), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    writer.Finalize(backingBuffer);
+
+    ReturnErrorOnFailure(mStorage->SyncSetKeyValue(GetStorageKey(node).KeyName(), backingBuffer.Get(), static_cast<uint16_t>(len)));
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR SimpleSubscriptionResumptionStorage::Save(SubscriptionInfo & subscriptionInfo)
+{
+    ScopedNodeId subscriptionNode = ScopedNodeId(subscriptionInfo.mNodeId, subscriptionInfo.mFabricIndex);
+    // Load index and update if fabric/node is new
+    SubscriptionIndex subscriptionIndex;
+    LoadIndex(subscriptionIndex);
+    bool nodeIsNew = true;
+    for (size_t i = 0; i < subscriptionIndex.mSize; i++)
+    {
+        if (subscriptionNode == subscriptionIndex.mNodes[i])
+        {
+            nodeIsNew = false;
+            break;
+        }
+    }
+    if (nodeIsNew)
+    {
+        if (subscriptionIndex.mSize == CHIP_IM_MAX_NUM_SUBSCRIPTIONS)
+        {
+            return CHIP_ERROR_NO_MEMORY;
+        }
+        subscriptionIndex.mNodes[subscriptionIndex.mSize++] = subscriptionNode;
+        SaveIndex(subscriptionIndex);
+    }
+
+    // Load existing subscriptions for node, then combine and save state
+    SubscriptionList subscriptions;
+    CHIP_ERROR err = FindByScopedNodeId(subscriptionNode, subscriptions);
```

Comment:
FindByScopedNodeId would return CHIP_NO_ERROR if the storage has no value for the storage key. In other error cases FindByScopedNodeId will delete the value and return error, and the current implementation is that the Save operation would fail as well. We still have time before release to evaluate if this is the right behavior or if the Save should be attempted anyway.

Reply:
No longer relevant because of storage scheme and API revamp to use iterator pattern.